### PR TITLE
feat: Add strip_top_level_identity_condition function

### PR DESCRIPTION
**Pull request by lynnagara**
Since Sentry currently uses the syntax `equals(or(x, y), 1)` as a
workaround for OR conditions we need to be able to strip the outer
equals condition in some cases and treat the nested boolean condition
as the top one.

A previous version of this PR changed the `get_first_level_and_conditions`
and `get_first_level_or_conditions` functions to do this automatically however
I don't think it's a good idea to apply this everywhere since those functions
are used widely, and the old behaviour would be expected in some cases.
For example it would cause the query formatting to be wrong.

Instead we add a new function that strips the outer identity condition which
will only be called in specific scenarios - such as in https://github.com/getsentry/snuba/pull/1289

### CODE REVIEW

File: snuba/query/conditions.py [hunk 1]
```diff
@@ -188,6 +191,27 @@ def _get_first_level_conditions(
...
+            FunctionCallPattern(
+                None, String(ConditionFunctions.EQ), (lit, nested_func_call),
+            ),
+        ]
+    )
+
+    match = id_pattern.match(condition)
+
+    if match:
+        condition = match.expression("nested_func")
```
Comment (fpacifici) [inline snuba/query/conditions.py:213]: This method is used in several key parts of query processing, like prewhere and project_id finder. Before making a functional change we need to ensure we are not breaking anything.
First I would suggest you run all sentry tests against snuba.
Also what about doing a dry run of this: match the condition and log if the result ids going to be different from the old implementation. In this case if nested_func.function_name is AND or OR, the result will be different.
Log these discrepancies in sentry, then we can have some confidence whether we are breaking anything.

**Review by fpacifici — CHANGES_REQUESTED** [review]

File: snuba/query/conditions.py [hunk 2]
```diff
@@ -188,6 +191,33 @@ def _get_first_level_conditions(
     In the AST, the condition is a tree, so we need some additional
     logic to extract the operands of the top level AND condition.
     """
+    # Ignore the identity pattern, e.g. cond = 1 or 1 = cond since Sentry uses this
+    # format frequently (in Discover) as a workaround to current schema limitations.
+    nested_func_call = Param(
+        "nested_func",
+        FunctionCallPattern(
+            None, Or([String(BooleanFunctions.AND), String(BooleanFunctions.OR)])
```
Comment (evanh) [inline snuba/query/conditions.py:199]: You _could_ make this more specific using `function`, e.g. `String(function)` but it's just a suggestion. I don't think it makes much difference besides removing the `Or`.

**Review by evanh — COMMENTED** [review]
LGTM, but I'll let Filippo give approval.

File: snuba/query/conditions.py (hunk 2 above)
Comment (lynnagara) [reply snuba/query/conditions.py:199]: Oh good point, thanks

**Comment by lynnagara** [comment]
After some more thought, I decided to totally change the approach here. Now we are not changing any of the existing functions only adding a new one that strips the outer condition if it matches the `equals(and(x, y), 1)` format. It's not called anywhere yet but will be used in https://github.com/getsentry/snuba/pull/1289

File: snuba/query/conditions.py (hunk 2 above)
Comment (evanh) [reply snuba/query/conditions.py:184]: On further reflection, you should move all these out as globals. There's no reason to build these matchers more than once.

**Review by evanh — APPROVED** [review]

**Comment by fpacifici** [comment]
Having two different methods around to decompose conditions, that do not work in the same way but such that, the one used widely is wrong and the correct one is only used in few cases does not seem a good idea. We are likely to keep these two around for a long time and, worse, write new features in an inconsistent way. 
Even worse, we will have to get rid of the wrong logic when it will have a much wider impact. The new condition is relatively new.

I bet the cases where the new logic (the correct tone) is behaving differently than the old one, are still very few. Could we still add a metric to confirm this ? If we have just a few cases I think we should just move to the new logic.

**Pull request closed without merging**